File: 2.OOP/concepts/MediumExample/design_patterns/Structural/Facade/Facade.js
// Light class representing a single lightbulb
class Light {
    constructor(name) {
      this.name = name;
    }
  
    turnOn() {
      console.log(`${this.name} light turned on`);
    }
  
    turnOff() {
      console.log(`${this.name} light turned off`);
    }
  }
  
  // LightSystem class representing the complex subsystem
  class LightSystem {
    constructor() {
      this.kitchenLight = new Light('Kitchen');
      this.livingRoomLight = new Light('Living Room');
    }
  
    turnOnKitchenLights() {
      this.kitchenLight.turnOn();
    }
  
    turnOffKitchenLights() {
      this.kitchenLight.turnOff();
    }
  
    turnOnLivingRoomLights() {
      this.livingRoomLight.turnOn();
    }
  
    turnOffLivingRoomLights() {
      this.livingRoomLight.turnOff();
    }
  }
  
  // LightController class acting as the facade
  class LightController {
    constructor() {
      this.lightSystem = new LightSystem();
    }
  
    turnOnAllLights() {
      this.lightSystem.turnOnKitchenLights();
      this.lightSystem.turnOnLivingRoomLights();
    }
  
    turnOffAllLights() {
      this.lightSystem.turnOffKitchenLights();
      this.lightSystem.turnOffLivingRoomLights();
    }
  }
  
  // Usage
  const lightController = new LightController();
  
  lightController.turnOnAllLights(); // Output: Kitchen light turned on, Living Room light turned on
  lightController.turnOffAllLights(); // Output: Kitchen light turned off, Living Room light turned off
  

File: 2.OOP/concepts/MediumExample/design_patterns/Structural/Facade/Facade.py
class Light:
  """Represents a single lightbulb."""
  def turn_on(self):
    print("Light turned on")

  def turn_off(self):
    print("Light turned off")

class LightSystem:
  """Represents the complex subsystem: the LightSystem."""
  def __init__(self):
    self.kitchen_light = Light()
    self.living_room_light = Light()

  def turn_on_kitchen_lights(self):
    self.kitchen_light.turn_on()

  def turn_off_kitchen_lights(self):
    self.kitchen_light.turn_off()

  def turn_on_living_room_lights(self):
    self.living_room_light.turn_on()

  def turn_off_living_room_lights(self):
    self.living_room_light.turn_off()

class LightController:
  """Facade class simplifying interaction with the LightSystem."""
  def __init__(self):
    self.light_system = LightSystem()

  def turn_on_all_lights(self):
    self.light_system.turn_on_kitchen_lights()
    self.light_system.turn_on_living_room_lights()

  def turn_off_all_lights(self):
    self.light_system.turn_off_kitchen_lights()
    self.light_system.turn_off_living_room_lights()

# Usage
light_controller = LightController()

light_controller.turn_on_all_lights()  # Output: Light turned on (twice)
light_controller.turn_off_all_lights()  # Output: Light turned off (twice)
